File: backend/app/schemas/regime.py
from typing import List, Optional
from pydantic import BaseModel


class MarkovKamaData(BaseModel):
    regime_code: List[int]                  # -2 to 2
    low_var_prob: List[Optional[float]]
    high_var_prob: List[Optional[float]]
    kama: List[Optional[float]]


class MSRegimeData(BaseModel):
    regime: List[int]                       # 0=low-stress, 1=high-stress
    regime_prob: List[Optional[float]]      # P(high-stress)


class YZPercentileData(BaseModel):
    yz_vol: List[Optional[float]]
    pct_rank: List[Optional[float]]         # 0–100


class TicaHmmData(BaseModel):
    regime_code:  List[int]   # 0..k-1 (Viterbi)
    regime_label: List[str]   # "Risk-On" | "Caution" | "Risk-Off"


class RegimeResponse(BaseModel):
    symbol: str
    timestamps: List[str]
    open: List[float]
    high: List[float]
    low: List[float]
    close: List[float]
    markov_kama: MarkovKamaData
    ms_regime: MSRegimeData
    yz_percentile: YZPercentileData
    tica_hmm: TicaHmmData


class RegimeRequest(BaseModel):
    start_date: Optional[str] = None
    end_date: Optional[str] = None

File: frontend/src/lib/services/regime.ts
import { API_BASE_URL } from '../api';
export interface MarkovKamaData {
  regime_code: number[];
  low_var_prob: (number | null)[];
  high_var_prob: (number | null)[];
  kama: (number | null)[];
}

export interface MSRegimeData {
  regime: number[];
  regime_prob: (number | null)[];
}

export interface YZPercentileData {
  yz_vol: (number | null)[];
  pct_rank: (number | null)[];
}

export interface TicaHmmData {
  regime_code:  number[];  // 0..k-1 Viterbi
  regime_label: string[];  // "Risk-On" | "Caution" | "Risk-Off"
}

export interface RegimeResponse {
  symbol: string;
  timestamps: string[];
  open: number[];
  high: number[];
  low: number[];
  close: number[];
  markov_kama: MarkovKamaData;
  ms_regime: MSRegimeData;
  yz_percentile: YZPercentileData;
  tica_hmm: TicaHmmData;
}

export interface RegimeRequest {
  start_date?: string;
  end_date?: string;
}

export const fetchRegime = async (
  symbol: string,
  params: RegimeRequest = {},
): Promise<RegimeResponse> => {
  const response = await fetch(
    `${API_BASE_URL}/regime/${symbol}`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(params),
    },
  );
  if (!response.ok) {
    throw new Error(`Error ${response.status}: ${await response.text()}`);
  }
  return response.json();
};
